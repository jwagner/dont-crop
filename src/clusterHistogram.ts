export function clusterHistogram(pixels: Float64Array, size: number, clusters: number) {
  const scaleAndClamp = (t: number) => (Math.max(0, Math.min(0.999, t)) * size) | 0;
  const cubeSize = size * size * size;
  const lcube = new Float64Array(cubeSize);
  const acube = new Float64Array(cubeSize);
  const bcube = new Float64Array(cubeSize);
  const ccube = new Float64Array(cubeSize);
  const lstep = 100 / size;
  for (let i = 0; i < pixels.length; i += 3) {
    const l = pixels[i];
    let a = pixels[i + 1];
    let b = pixels[i + 2];
    if (l < lstep || l > 100 - lstep) {
      a = 0;
      b = 0;
    }
    const x = scaleAndClamp(l / 100);
    const y = scaleAndClamp((a + 100) / 200);
    const z = scaleAndClamp((b + 100) / 200);
    const j = x + y * size + z * size * size;
    lcube[j] += l;
    acube[j] += a;
    bcube[j] += b;
    ccube[j]++;
  }
  return Array.from(lcube.keys())
    .filter((i) => ccube[i])
    .sort((i, j) => ccube[j] - ccube[i])
    .slice(0, clusters)
    .map((i) => ({
      x: lcube[i] / ccube[i],
      y: acube[i] / ccube[i],
      z: bcube[i] / ccube[i],
      count: ccube[i],
    }));
}
