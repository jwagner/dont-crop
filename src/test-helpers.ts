import { join } from 'path';
import sharp from 'sharp';
import { srgbToLabComposed } from './lab';

export async function getTestImageData(): Promise<ImageData> {
  const src = join(__dirname, '../tests/image.jpg');
  const { data, info } = await sharp(src).ensureAlpha()
    .toColorspace('srgb')
    .raw()
    .toBuffer({ resolveWithObject: true });
  const imageData: ImageData = {
    width: info.width,
    height: info.height,
    data: new Uint8ClampedArray(data.buffer),
  };
  return imageData;
}

export async function getLabTestData() {
  const imageData = await getTestImageData();
  return srgbToLabComposed(imageData.data);
}
