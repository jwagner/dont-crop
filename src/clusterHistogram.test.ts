import { clusterHistogram } from './clusterHistogram';
import { getLabTestData } from './test-helpers';

describe('clusterHistogram', () => {
  test('handles an empty input', () => {
    const actual = clusterHistogram(new Float64Array(0), 6, 1);
    expect(actual).toHaveLength(0);
  });
  test('handles a single input', () => {
    const expected = {
      x: 32,
      y: -16,
      z: 64,
      count: 1,
    };
    const pixels = new Float64Array([expected.x, expected.y, expected.z]);
    const actual = clusterHistogram(pixels, 6, 1);
    expect(actual).toEqual([expected]);
  });
  test('matches snapshot', async () => {
    const pixels = await getLabTestData();
    const actual = clusterHistogram(pixels, 6, 4);
    expect(actual).toMatchSnapshot();
  });
});
