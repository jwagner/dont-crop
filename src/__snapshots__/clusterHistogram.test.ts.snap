// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`clusterHistogram matches snapshot 1`] = `
[
  {
    "count": 1877,
    "x": 5.450388706434306,
    "y": 0,
    "z": 0,
  },
  {
    "count": 1297,
    "x": 74.19850345616341,
    "y": 17.59626962059904,
    "z": 50.07157975362288,
  },
  {
    "count": 1082,
    "x": 89.07517793139733,
    "y": 0,
    "z": 0,
  },
  {
    "count": 857,
    "x": 58.67642932627202,
    "y": 46.0839467883922,
    "z": 50.218335296709554,
  },
]
`;
